src/main/java/org/takes/misc/Transform.java

```
@@ -0,0 +1,112 @@
+ *
+ * @author Jason Wong (super132j@yahoo.com)
+ * @version $Id$
+ * @since 0.13.8
+ */
+public class Transform<T, K> implements Iterable<K> {
+
+    /**
+     * Internal storage.
+     */
+    private final transient Iterable<T> list;
+
+    /**
+     * The action to transform the elements in the iterator.
+     */
+    private final transient TransformAction<T, K> action;
+
+    /**
+     * Transform elements in the supplied iterable by the action supplied.
+     * @param itb Iterable to be transformed
+     * @param act The actual transformation implementation
+     */
+    public Transform(final Iterable<T> itb,
+            final TransformAction<T, K> act) {
+        this.list = itb;
+        this.action = act;
+    }
+
+    @Override
+    public final Iterator<K> iterator() {
+        return new TransformIterator<T, K>(this.list.iterator(), this.action);
+    }
+
+    /**
+     * The iterator to iterator through the original type B and return the
+     * transformed element in type A.
+     */
+    private static class TransformIterator<B, A> implements Iterator<A> {
+
+        /**
+         * The iterator to reflect the traverse state.
+         */
+        private final transient Iterator<B> iterator;
+
+        /**
+         * The action to transform the elements in the iterator.
+         */
+        private final transient TransformAction<B, A> action;
+
+        /**
+         * Ctor. ConcatIterator traverses the element.
+         * @param itr Iterator of the original iterable
+         * @param act Action to transform elements
+         */
+        public TransformIterator(final Iterator<B> itr,
+                final TransformAction<B, A> act) {
+            this.action = act;
+            this.iterator = itr;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return this.iterator.hasNext();
+        }
+
+        @Override
+        public A next() {
+            if (this.hasNext()) {
```

I don't understand why do we need to call `hasNext()` here? we can just use `this.iterator.next()` and let it throw if there is nothing left